packages/web-components/fast-element/src/styles/element-styles.ts

```diff
@@ -44,34 +37,101 @@ export interface StyleTarget {
 export type ComposableStyles = string | ElementStyles | CSSStyleSheet;
 
 /**
- * Creates an ElementStyles instance for an array of ComposableStyles.
+ * Implemented to provide specific behavior when adding/removing styles
+ * for elements.
  * @public
  */
-export type ElementStyleFactory = (
+export interface StyleStrategy {
+    /**
+     * Adds styles to the target.
+     * @param target - The target to add the styles to.
+     */
+    addStylesTo(target: StyleTarget): void;
+
+    /**
+     * Removes styles from the target.
+     * @param target - The target to remove the styles from.
+     */
+    removeStylesFrom(target: StyleTarget): void;
+}
+
+/**
+ * A type that instantiates a StyleStrategy.
+ * @public
+ */
+export type ConstructibleStyleStrategy = {
+    /**
+     * Creates an instance of the strategy.
+     * @param styles - The styles to initialize the strategy with.
+     */
+    new (styles: (string | CSSStyleSheet)[]): StyleStrategy;
+};
+
+const styleSheetCache = new Map<string, CSSStyleSheet>();
+let DefaultStyleStrategy: ConstructibleStyleStrategy;
+
+function reduceStyles(
     styles: ReadonlyArray<ComposableStyles>
-) => ElementStyles;
+): (string | CSSStyleSheet)[] {
+    return styles
+        .map((x: ComposableStyles) =>
+            x instanceof ElementStyles ? reduceStyles(x.styles) : [x]
+        )
+        .reduce((prev: string[], curr: string[]) => prev.concat(curr), []);
+}
 
 /**
  * Represents styles that can be applied to a custom element.
  * @public
  */
-export abstract class ElementStyles {
+export class ElementStyles {
     private targets: WeakSet<StyleTarget> = new WeakSet();
+    private _strategy: StyleStrategy | null = null;
 
-    constructor(
-        /** @internal */
-        public readonly styles: ReadonlyArray<ComposableStyles>,
+    /** @internal */
+    public readonly behaviors: ReadonlyArray<Behavior<HTMLElement>> | null;
+
+    /**
+     * Gets the StyleStrategy associated with these element styles.
+     */
+    public get strategy(): StyleStrategy {
```

Comment:
Using this property, SSR can retrieve the strategy used by an element's styles. SSR would provide its own default strategy type (see below) and so it can then cast to that type and retrieve whatever information it may need.